src/log/log_slot.c

```diff
@@ -30,12 +72,25 @@ __wt_log_slot_activate(WT_SESSION_IMPL *session, WT_LOGSLOT *slot)
 	 * set for closing the file handle on a log file switch.  The flags
 	 * are reset when the slot is freed.  See log_slot_free.
 	 */
+	slot->slot_unbuffered = 0;
 	slot->slot_start_lsn = slot->slot_end_lsn = log->alloc_lsn;
 	slot->slot_start_offset = log->alloc_lsn.l.offset;
 	slot->slot_last_offset = log->alloc_lsn.l.offset;
 	slot->slot_fh = log->log_fh;
 	slot->slot_error = 0;
-	slot->slot_unbuffered = 0;
+#ifdef	HAVE_DIAGNOSTIC
+	/*
+	 * Yield here to encourage races during slot switches.
+	 */
+	__wt_yield();
+#endif
+	WT_BARRIER();
+	/*
+	 * Set the slot state last.  Other threads may have a stale pointer
+	 * to this slot everything else needs to be there before the state
+	 * is cleared.  Use a barrier to make sure it is last.
+	 */
+	slot->slot_state = 0;
```

Comment:
It isn't clear to me that the `WT_BARRIER` above guarantees anything in terms of correctness.  (I suspect the compiler is generating code in the order specified here in the absence of any reason to reorder, but the `WT_BARRIER` doesn't guarantee that cores will see writes in the order they are issued).

Do you mean `WT_PUBLISH(slot->slot_state, 0);`?